guacamole/src/main/java/org/apache/guacamole/tunnel/StandardTokenMap.java

```diff
@@ -102,6 +115,13 @@ public StandardTokenMap(AuthenticatedUser authenticatedUser) {
         else
             put(USERNAME_TOKEN, authenticatedUser.getIdentifier());
 
+        if (get(USERNAME_TOKEN).contains("\\")) {
+            put(USERNAME_DOMAIN_TOKEN, get(USERNAME_TOKEN).split("\\\\")[0]);
+            put(USERNAME_ID_TOKEN, get(USERNAME_TOKEN).split("\\\\")[1]);
+        } else {
+            put(USERNAME_DOMAIN_TOKEN, "");
+            put(USERNAME_ID_TOKEN, get(USERNAME_TOKEN));
+        }
```

Comment:
I don't think decisions should be made about the format of the username at the web application level, nor that we should make specific assumptions about the format here when the LDAP handling of username format is so flexible.

What about exposing `authenticatedUser.getIdentifier()` as something like `GUAC_IDENTIFIER`? That would allow the unique identifier that the LDAP support is already parsing out of the username to be exposed automatically without forcing the webapp to be aware of LDAP-specific details.

If it is also necessary to pull the domain of a user authenticating with LDAP and make that available with a token, it would make sense to do that within the LDAP support, and it may be possible to extend the way `match-usernames` works to provide that.

Reply:
> I don't think decisions should be made about the format of the username at the web application level...

I agree, and my question would be: Why do these need to be standard tokens at all? We've provided a mechanism within the LDAP module (and most of the other modules, particularly SSO) to define tokens specific to those modules, but that should be usable in connections stored in other modules. Is there any reason not to just do something like:
* LDAP_USER_REALM or LDAP_USER_DOMAIN
* LDAP_USER_IDENTIFIER, or just pull the LDAP attribute with the username?

? Or is there some other piece of this I'm missing?

Reply:
If there is to be a new standard token, I think there's some value in having one token (`GUAC_IDENTIFIER`? `GUAC_USER_IDENTIFIER`?) that represents validated identity while the other (`GUAC_USERNAME`) represents the username provided during the auth process, if any.

I'd also have no issue if this were 100% purely LDAP-specific tokens.

> ... or just pull the LDAP attribute with the username?

I think that much is already possible with the functionality provided by the `ldap-user-attributes` property.

Reply:
> If there is to be a new standard token, I think there's some value in having one token (GUAC_IDENTIFIER? GUAC_USER_IDENTIFIER?) that represents validated identity while the other (GUAC_USERNAME) represents the username provided during the auth process, if any.

Sure, that makes sense - it doesn't resolve the issue of getting the domain portion out of the overall username, so if there's a need to have the domain, that would still have to be handled - which seems more appropriate to be LDAP-specific.

> I think that much is already possible with the functionality provided by the ldap-user-attributes property.

Yeah, I was looking at that code, but also did a quick search of my local AD environment and figured out that the domain is not (obviously) present in any attribute without splitting it out of another attribute. So, again, if the domain portion is desired/required, it would need to be something LDAP-specific.

If the desire is just to get the Guacamole identifier, I'm good with the `GUAC_IDENTIFIER` route.

Reply:
`if the domain portion is desired/required, it would need to be something LDAP-specific` - Yes these variables we are looking are LDAP-specific. `authenticatedUser.getIdentifier()` only gives us which authentication mechanism i have used for my guacamole instance. 
Also In the instance multiple LDAP configurations with `match-usernames` `${GUAC_USERNAME` becomes `domain\username` which is not ideal for RDP connections.  Check below some parts error 
```
    Failure Reason:     Unknown user name or bad password.
    Status:         0xC000006D
    Sub Status:     0xC0000064
 
Process Information:
    Caller Process ID:  0x0
    Caller Process Name:    -

Detailed Authentication Information:
    Logon Process:      NtLmSsp
    Authentication Package: NTLM
    Transited Services: -
    Package Name (NTLM only):   -
    Key Length:     0
```

It would be ideal to have DOMAIN and just user name which are again LDAP sepcific, it makes easy for connection configuration without any hassle.

Reply:
> `authenticatedUser.getIdentifier()` only gives us which authentication mechanism i have used for my guacamole instance.

What do you mean? The `AuthenticatedUser` is the object defining the user's authenticated identity, with `getIdentifier()` returning the unique identifier for that identity. In the case of LDAP, that is the username. This is exactly the value that is returned as the username to be exposed in the UI:

https://github.com/apache/guacamole-client/blob/284d1b24f37f06d380312e149d24d96497dd9af4/guacamole/src/main/java/org/apache/guacamole/rest/auth/TokenRESTService.java#L188C1-L194

Providing the identifier as a token would allow the relevant username to be included in the configuration independently of what the user enters in a login form.

Reply:
You're right @necouchman , main intent is to expose a token (agree should be LDAP specific) containing only the domain (and another containing only the username) that can be used in connection definitions.

> We've provided a mechanism within the LDAP module (and most of the other modules, particularly SSO) to define tokens specific to those modules, but that should be usable in connections stored in other modules.

To do that, I presume we can add tokens to what's already being collected here?https://github.com/apache/guacamole-client/blob/284d1b24f37f06d380312e149d24d96497dd9af4/extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/AuthenticationProviderService.java#L355-L358

Reply:
@fanovilla Yes, that function (`getAttributeTokens()`) is a reasonable place to add LDAP-specific tokens. If you do so, the name and documentation for that function would need to be adjusted, as it currently states:

> Returns parameter tokens generated from LDAP attributes on the user currently bound under the given LDAP connection. ...

If a standard token for the user identifier is being added, as well, `StandardTokenMap` is correct for that.

Reply:
thanks @mike-jumper ; @josnabattula not sure how but let me know if you want me to try contribute to your PR branch..

Reply:
Yep, you can.